### PR TITLE
Adding new debug tools to the scene Inspector : copyCommandToClipboar…

### PR DESCRIPTION
**Pull request by Tricotou**
Same version than final state of [previous PR 15050](https://github.com/BabylonJS/Babylon.js/pull/15050/)
  PLUS the following :
____
Added a fix (`.SCSS` only) on some broken tools, NME, NGE, gui-editor, where copy buttons where appearing, due to modification of the sharedUI line components.
____
Built locally and tested on `Playground`, `Sandbox`, `NME`, `NGE`, `GuiEditor`

### CODE REVIEW

**Comment by bjsplat** [comment]
Please make sure to label your PR with "bug", "new feature" or "breaking change" label(s).
To prevent this PR from going to the changelog marked it with the "skip changelog" label.

**Comment by deltakosh** [comment]
Ok let's test on the CI version of the tools and we should be good to go!

**Comment by bjsplat** [comment]
Snapshot stored with reference name:
refs/pull/15066/merge

Test environment: 
https://babylonsnapshots.z22.web.core.windows.net/refs/pull/15066/merge/index.html

To test a playground add it to the URL, for example:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/15066/merge/index.html#WGZLGJ#4600

Links to test babylon tools with this snapshot:

https://playground.babylonjs.com/?snapshot=refs/pull/15066/merge
https://sandbox.babylonjs.com/?snapshot=refs/pull/15066/merge
https://gui.babylonjs.com/?snapshot=refs/pull/15066/merge
https://nme.babylonjs.com/?snapshot=refs/pull/15066/merge

To test the snapshot in the playground with a playground ID add it after the snapshot query string:

https://playground.babylonjs.com/?snapshot=refs/pull/15066/merge#BCU1XR#0

**Comment by bjsplat** [comment]
WebGL2 visualization test reporter:

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/15066/merge/testResults/webgl2playwright/index.html

**Comment by Tricotou** [comment]
I'm sorry it's still not good. Don't merge this snapshot :)

**Comment by deltakosh** [comment]
seems to still be broken on nme
image

**Comment by deltakosh** [comment]
https://nme.babylonjs.com/?snapshot=refs/pull/15066/merge

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/15066/merge/testResults/webgpuplaywright/index.html

**Comment by Tricotou** [comment]
Ok so here is the status to me :
* In this new PR I had commited a test file on the playground, sorry for this un neat stuff (it's reverted)
* About the tests on the tools, I don't get why the version of the CI is broken. I just tested theses :
```
npm run serve -w @tools/babylon-server
```
And in another terminal, the one after another :
```
npm run serve -w @tools/node-editor
npm run serve -w @tools/node-geometry-editor
npm run serve -w @tools/gui-editor
```
Each of them appear fine :
Screenshot from 2024-05-04 00-40-42
Screenshot from 2024-05-04 00-41-26
Screenshot from 2024-05-04 00-42-08

I must be missing something with the version running under the `npm run serve` command... :thinking:

**Comment by deltakosh** [comment]
Let's wait for the build grandmaster to help:) @RaananW

**Comment by RaananW** [comment]
When opening the tools with snapshots they use the core libraries from the PR, but still use the tools packages from master. Therefore, you might still be seeing errors. The best way to test the tools from a snapshot is opening them from the inspector. For example, open the playground's snapshot , open an NME playground, and press the edit button. It will then work as expected:

image

**Review by RaananW — DISMISSED** [review]
I find this feature very helpful, however I have just one little request. When copying a value you get something like this:

`BABYLON.Mesh.scaling = new BABYLON.Vector3(1, 1, 1);` 

This is quite confusing - copying this to the console will actually "work", but it will do nothing. I would expect to copy the values, and not the class itself, or, even if you want to expose the class itself:

1) Not assign it as a static variable to the global namespace (or not assign it at all)
2) Not use the global BABYLON namespace automatically, which is UMD-only.

I believe that when copying a vector I should get `[x,y,z]`, or maybe a `meshScaling.set(x,y,z)` if we want to have assignment. The BABYLON namespace should only be used if it exists, but I wouldn't want to add inspection for UMD or global namespace in the core library.
I will be happy to discuss this, or to be convinced I am wrong :-)

**Comment by Tricotou** [comment]
Hello Raanan, thanks for your review ! :)

I agree it can be confusing. The copied command `BABYLON.Mesh.scaling = new BABYLON.Vector3(1, 1, 1);` is not meant to be executed "as is", to me. Maybe it could be done differently, but here is my original idea :
* I wanted to have a way to retreive the value (right part)
* I wanted to explicit the type Class (and arborescence) of the property (left part)

Following my original [post](https://forum.babylonjs.com/t/scene-debuglayer-copy-last-command-to-clipboard/49796) for example the "Environment" param inside the "Level" tab of material, was confusing since the param is not `Material.Environment.Level`, but `material.environmentIntensity`. So that's why my idea was to explicit the `material` to be a `BABYLON.Material` object resulting in the `BABYLON.Material.environmentIntensity = 0.5;` line :)
____
Maybe we could copy this line : (including comment)
```javascript
mesh.scaling = new BABYLON.Vector3(1, 1, 1);// mesh as BABYLON.Mesh
```

What do you think ?

**Comment by RaananW** [comment]
> Maybe we could copy this line : (including comment)
> 
> ```js
> mesh.scaling = new BABYLON.Vector3(1, 1, 1);// mesh as BABYLON.Mesh
> ```

I am fine with this change. I am not sure we need to create a new vector in this case, and `set` will make more sense, but that would require you knowing what class is being used. The only thing I still have a problem with is the BABYLON namespace, which will be wrong in the es6 packages. Not sure if there is a solution for this TBH. I guess, since the dev already expected to know that the assignment might fail, removing the BABYLON namespace can make sense as well - expect the user to be sure how to construct the class they are trying to set.

**Comment by RaananW** [comment]
> To be perfectly "neat" cannot we detect which packages are in use, and eventually automatically add the `BABYLON` namespace if required ?

I am not a big fan of detecting this in code, but as long as it is not in core, you can check if the namespace exists (i.e. if typeof BABYLON !== "undefined") and decide whether or not to use it. As long as it is tools-only feature.



> And what do you think about the second idea ?

I am not sure where `debugNode` comes from (I am probably missing something :-) ). I think the first suggestion is better, as it forces a difference between two nodes.

**Comment by Tricotou** [comment]
> as long as it is not in core

By "core" do you mean only `/dev/core/` or the whole `/dev/` folder ?
(Here we are speaking about some code inside `/dev/sharedUIComponents/`)

> I am probably missing something :-)

This feature is part of the same PR. See the [post 18 of original topic ](https://forum.babylonjs.com/t/scene-debuglayer-copy-last-command-to-clipboard/49796/18) or the [first comment of the original PR](https://github.com/BabylonJS/Babylon.js/pull/15050/). Basically, `window.debugNode` is set as reference toward the selected node in the Inspector hierachy. So that you can then use it directly in the console :
* You select a mesh and type `let mesh = window.debugNode;`
* You select a material and type `let mat = window.debugNode;`
And then you can work in the console using `mesh` and `mat` :)

**Comment by RaananW** [comment]
> /dev/sharedUIComponents/

That's fine, I am more referring to the basic packages like @babylonjs/core and loaders etc'.

> `window.debugNode` is set as reference toward the selected node in the Inspector hierachy

As long as this is inspector code, this is probably fine. Would be better to use `globalThis` instead of window just to be safe.
So I am fine with the copy to window.debugNode, though I am not sure what it will bring to the person who copied it, as this is already the value of the debugNode directly, no? Or is the idea to go to a different node, and use the copy from another node?

Oh, and one last thing - you will need to make sure to clear the debugNode from the `globalThis` object to avoid memory leaks.

**Comment by Tricotou** [comment]
> That's fine, I am more referring to the basic packages like @babylonjs/core and loaders etc'.

Ok

> As long as this is inspector code, this is probably fine.

Ok

> though I am not sure what it will bring to the person who copied it, as this is already the value of the debugNode directly, no? Or is the idea to go to a different node, and use the copy from another node?

To me the idea is only to retrieve a pointer toward the selected node, in `window.debugNode` (or `globalThis.debugNode`).
So each time the user selects another node, the var `globalThis.debugNode` would only be pointing toward another object, but without any cost in memory, or object copy or anything.

Then, if the user does `const mesh = globalThis.debugNode;` it's just a way to keep aside the reference to this specific mesh, before eventually clicking another one.

> you will need to make sure to clear the debugNode from the `globalThis` object to avoid memory leaks.

Ok I'll have a look at this :slightly_smiling_face:

**Comment by Tricotou** [comment]
New version has been pushed including :

* Command copied is now executable "as is"
* `window.debugNode` :arrow_right: `globalThis.debugNode`
* `globalThis.debugNode` deleted in the `scene.onDisposeObservable` of the scene attached to the node (GC)
* `BABYLON` namespace is removed from command if undefined (ES6 support)
____
`NB`: This new PR now includes the fix on the `.scss` files of the tools previously broken (NME, NGE, GUI)
But if I understood well, it's still possible that after the merge, online version would be broken for a few time until both CI pipeline are synced (Tools + Core).

@RaananW do you have a way to force an update on the tools pipelines at the same time, in order to make sure no tool remains broken ?

**Comment by RaananW** [comment]
Not from a PR, but I can check all tools from the playground. I'll do that very soon

**Comment by RaananW** [comment]
NME still has this issue:

image

Couldn't find any issue with NGE or the GUI editor

**Comment by Tricotou** [comment]
Hum. It's weird because this copy button appears within a `.listLine` class (As you can see it's just under the Blender mode param, with a dropdown menu) and it's fixed in the `.scss` file. We can see the fix in the same panel on the top (ex : Alpha mode) there are some other listLine without the problem :thinking: 

Let me check the code again, I might me missing another .scss which is overrinding this part (preview config bar)

**Comment by bjsplat** [comment]
Visualization tests for WebGPU (Experimental)
Important - these might fail sporadically. This is an optional test.

https://babylonsnapshots.z22.web.core.windows.net/refs/pull/15066/merge/testResults/webgpuplaywright/index.html

**Comment by Tricotou** [comment]
> NME still has this issue

Fixed ✅

**Pull request merged**
